pkg/rules/proxy.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package rules
+
+import (
+	"context"
+	"io"
+	"sort"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/rules/rulespb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Proxy implements rulespb.Rules that fanouts requests to given rulespb.Rules and deduplication on the way.
+type Proxy struct {
+	logger log.Logger
+	rules  func() []rulespb.RulesClient
+}
+
+// NewProxy returns new rules.Proxy.
+func NewProxy(logger log.Logger, rules func() []rulespb.RulesClient) *Proxy {
```

Comment:
Minor thing but the StoreAPI method is called `NewProxyStore` so maybe let's call this `NewProxyRule`?

Reply:
yup

Reply:
oh actually it will be done on next PR

Reply:
actually let's leave it Proxy. Store thing should be fixed IMO